Optimize HTML parser entity names table by omitting semicolons

darinadler opened the pull request:
#### 3483dcf98d883183eb0621479ed8f19451533722
<pre>
Optimize HTML parser entity names table by omitting semicolons
<a href="https://bugs.webkit.org/show_bug.cgi?id=253358">https://bugs.webkit.org/show_bug.cgi?id=253358</a>
rdar://problem/106264459

Reviewed by Chris Dumez.

Crunched the table further using bitfields and omitting semicolons.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Added create-html-entity-table so searches will see source code there.

* Source/WebCore/html/parser/HTMLEntityParser.cpp:
(WebCore::HTMLEntityParser::consumeNamedEntity): Updated for the changes
to HTMLEntityTableEntry.
(WebCore::appendUChar32ToUCharArray): Deleted.
(WebCore::decodeNamedEntityToUCharArray): Use U16_APPEND_UNSAFE instead of
our own appendUChar32ToUCharArray.

* Source/WebCore/html/parser/HTMLEntitySearch.cpp:
(WebCore::midpoint): Renamed from halfway.
(WebCore::HTMLEntitySearch::compare const): Added code that can compare
with the semicolon even though it&apos;s not actually stored in the HTML entity table.

* Source/WebCore/html/parser/HTMLEntitySearch.h: Tweaked coding style
and names a bit.

* Source/WebCore/html/parser/HTMLEntityTable.h: Renamed all the members of
HTMLEntityTableEntry for clarity. There are two sets of characters, the name
of the entity, and the expansion of the entity, and the names must disambiguate.
The new entries fit in 57 bits with bit fields. Possible could crunc more if we
could save one more bit.

* Source/WebCore/html/parser/create-html-entity-table:
Write out the new table fields, omit semicolons. Check values fit in bit fields.

Canonical link: <a href="https://commits.webkit.org/261734@main">https://commits.webkit.org/261734@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/28fa26ef4f27e944239936f060acc0709ed79617

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112690 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21846 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1362 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4466 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121216 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116758 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23187 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13011 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/5618 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118459 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100460 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105769 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/1004 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46242 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14167 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1042 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14854 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20184 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53037 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8186 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16698 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/parser/create-html-entity-table:
```diff
@@ -91,47 +91,39 @@ output_file.write("""/*
     entity = entry[ENTITY]
+    check_ascii(entity)
+    if entity[-1] == ';':
+        entity = entity[0:-1]
```
nit: `entity = entity[:-1]` would also work and slightly more concise.

darinadler replied:
OK, done.

cdumez commented on Source/WebCore/html/parser/create-html-entity-table:
```diff
@@ -161,47 +152,43 @@ for offset, entry in enumerate(entries):
+    name_includes_trailing_semicolon = int(entry[ENTITY][-1] == ';')
+    if name_includes_trailing_semicolon:
+        name_length -= 1
+    assert first_character < (1 << 21)
```
May be nice to have a comment referring to the header where those bitfields are defined, to make sure we know where those values come from and make sure they stay in sync.

darinadler replied:
Agreed. Did that.

cdumez reviewed (approved):
LGTM

darinadler committed:
Optimize HTML parser entity names table by omitting semicolons
https://bugs.webkit.org/show_bug.cgi?id=253358
rdar://problem/106264459

Reviewed by Chris Dumez.

Crunched the table further using bitfields and omitting semicolons.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Added create-html-entity-table so searches will see source code there.

* Source/WebCore/html/parser/HTMLEntityParser.cpp:
(WebCore::HTMLEntityParser::consumeNamedEntity): Updated for the changes
to HTMLEntityTableEntry.
(WebCore::appendUChar32ToUCharArray): Deleted.
(WebCore::decodeNamedEntityToUCharArray): Use U16_APPEND_UNSAFE instead of
our own appendUChar32ToUCharArray.

* Source/WebCore/html/parser/HTMLEntitySearch.cpp:
(WebCore::midpoint): Renamed from halfway.
(WebCore::HTMLEntitySearch::compare const): Added code that can compare
with the semicolon even though it's not actually stored in the HTML entity table.

* Source/WebCore/html/parser/HTMLEntitySearch.h: Tweaked coding style
and names a bit.

* Source/WebCore/html/parser/HTMLEntityTable.h: Renamed all the members of
HTMLEntityTableEntry for clarity. There are two sets of characters, the name
of the entity, and the expansion of the entity, and the names must disambiguate.
The new entries fit in 57 bits with bit fields. Possible could crunc more if we
could save one more bit.

* Source/WebCore/html/parser/create-html-entity-table:
Write out the new table fields, omit semicolons. Check values fit in bit fields.

Canonical link: https://commits.webkit.org/261734@main
